azurelinuxagent/common/utils/processutil.py

```diff
@@ -0,0 +1,226 @@
+#
+
+import subprocess
+import sys
+import os
+import time
+import signal
+from errno import ESRCH
+from multiprocessing import Process
+
+from azurelinuxagent.common.exception import ExtensionError
+from azurelinuxagent.common.future import ustr
+
+TELEMETRY_MESSAGE_MAX_LEN = 3200
+
+
+def sanitize(s):
+    return ustr(s, encoding='utf-8', errors='backslashreplace')
+
+
+def format_stdout_stderr(stdout, stderr, max_len=TELEMETRY_MESSAGE_MAX_LEN):
+    """
+    Format stdout and stderr's output to make it suitable in telemetry.
+    The goal is to maximize the amount of output given the constraints
+    of telemetry.
+
+    For example, if there is more stderr output than stdout output give
+    more buffer space to stderr.
+
+    :param str stdout: characters captured from stdout
+    :param str stderr: characters captured from stderr
+    :param int max_len: maximum length of the string to return
+
+    :return: a string formatted with stdout and stderr that is less than
+    or equal to max_len.
+    :rtype: str
+    """
+    template = "[stdout]\n{0}\n\n[stderr]\n{1}"
+    # +6 == len("{0}") + len("{1}")
+    max_len_each = int((max_len - len(template) + 6) / 2)
+
+    if max_len_each <= 0:
+        return ''
+
+    def to_s(captured_stdout, stdout_offset, captured_stderr, stderr_offset):
+        s = template.format(captured_stdout[stdout_offset:], captured_stderr[stderr_offset:])
+        return s
+
+    if len(stdout) + len(stderr) < max_len:
+        return to_s(stdout, 0, stderr, 0)
+    elif len(stdout) < max_len_each:
+        bonus = max_len_each - len(stdout)
+        stderr_len = min(max_len_each + bonus, len(stderr))
+        return to_s(stdout, 0, stderr, -1*stderr_len)
+    elif len(stderr) < max_len_each:
+        bonus = max_len_each - len(stderr)
+        stdout_len = min(max_len_each + bonus, len(stdout))
+        return to_s(stdout, -1*stdout_len, stderr, 0)
+    else:
+        return to_s(stdout, -1*max_len_each, stderr, -1*max_len_each)
+
+
+def _destroy_process(process, signal_to_send=signal.SIGKILL):
+    """
+    Completely destroy the target process. Close the stdout/stderr pipes, kill the process, reap the zombie.
+    If process is the leader of a process group, kill the entire process group.
+
+    :param Popen process: Process to be sent a signal
+    :param int signal_to_send: Signal number to be sent
+    """
+    process.stdout.close()
+    process.stderr.close()
+    try:
+        pid = process.pid
+        if os.getpgid(pid) == pid:
+            os.killpg(pid, signal_to_send)
+        else:
+            os.kill(pid, signal_to_send)
+        process.wait()
+    except OSError as e:
+        if e.errno != ESRCH:
+            raise
+        pass    # If the process is already gone, that's fine
+
+
+def capture_from_process_modern(process, cmd, timeout):
+    try:
+        stdout, stderr = process.communicate(timeout=timeout)
+    except subprocess.TimeoutExpired:
+        os.killpg(os.getpgid(process.pid), 9)
```

Comment:
You don't call _destroy_process because it closes the STDOUT/STDERR and you need to call .communicate() on more time?

Reply:
Exactly correct. .communicate() will read stdout/stderr until EOF and then close them; .communicate() will then .wait() on the process, reaping the zombie. I need do nothing more than kill the process, then let .communicate() finish its job.

Reply:
I would add a comment.